docs/development/pr-review.asciidoc

```diff
@@ -0,0 +1,129 @@
++
+--
+Making poor decisions around how a change is technically architected means we immediately have taken on technical debt, which is going to unnecessarily distract us from our future efforts to improve the project as we need to maintain that existing change, chase down bugs, and ultimately re-architect that area in the future.
+
+There is no one-size-fits-all way to evaluate whether a given change is technically sound, so instead we must rely on our concrete knowledge and experience building software. This is an especially hard thing to do when two experienced developers have differing opinions about what is and is not architecturally sound, so it requires both the author and reviewer to show empathy, to go out of their way to understand each other, and to come to a consensus about a path forward.
+
+The goal here isn't finding the one best architecture as it's unlikely this could ever be qualified. There are usually many architecturally sound solutions to any problem, so reviewers should be focussed on that goal objectively rather than whether or not their own vision for how the change should be architected is "better".
+--
+. *Are there sufficient tests to ensure this change doesn't break in the future?*
++
+--
+We cannot scale our development efforts and continue to be effective at improving and maintaining the product if we're relying on manual testing to catch bugs. Manual testing is our last line of defense, and it's the least effective and most expensive way to ensure a stable product. This means that every change we make to the product needs to have sufficient test coverage.
+
+This isn't simply a question of enough test files. The code in the tests themselves is just as important as the code in the product.
+
+All of our code should have unit tests that verify its behaviors, including not only the "happy path", but also edge cases, error handling, etc.  When you change an existing API of a module, then there should always be at least one failing unit test, which in turn means we need to verify that all code consuming that API properly handles the change if necessary. For modules at a high enough level, this will mean we have breaking change in the product, which we'll need to handle accordingly.
+
+In addition to extensive unit test coverage, PRs should include relevant functional and integration tests. In some cases, we may simply be testing a programmatic interface (e.g. a service) that is integrating with the file system, the network, Elasticsearch, etc. In other cases, we'll be testing REST APIs over HTTP or comparing screenshots/snapshots with prior known acceptable state. In the worst case, we are doing browser-based functional testing on a running instance of Kibana using selenium.
+
+Enhancements are pretty much always going to have extensive unit tests as a base as well as functional and integration testing. Bug fixes should always include regression tests to ensure that same bug does not manifest again in the future.
```

Comment:
Not necessarily relevant for this PR itself, but at the same time it kinda is, so I'm including it: I wish we had a "testing doc" in Kibana. Something that linked to examples of good tests of different types and talked about testing strategy more "practically". This is okey as a start, but it's not that easy (even for me) to understand our test setup. E.g. every time I'm using the functional test runner to run api tests I need to use my bash history to even find the command (which I found searching for the folder on github: https://github.com/elastic/kibana/search?utf8=%E2%9C%93&q=api_integration). We can do a much better job here, imo.

Reply:
I couldn't agree more. Whenever we do, we should update this section to link to it as well.